### PR TITLE
Add ADF4382

### PR DESCRIPTION
**Pull request by chegbeli**
## Pull Request Description

The ADF4382A is a high performance, ultralow jitter, Frac-N PLL
with integrated VCO ideally suited for LO generation for 5G applications
or data converter clock applications. The high performance
PLL has a figure of merit of -239 dBc/Hz, low 1/f Noise and
high PFD frequency of 625MHz in integer mode that can achieve
ultralow in-band noise and integrated jitter. The ADF4382A can
generate frequencies in a fundamental octave range of 11.5 GHz to
21 GHz, thereby eliminating the need for sub-harmonic filters. The
divide by 2 and 4 output dividers on the part allow frequencies to
be generated from 5.75GHz to 10.5GHz and 2.875GHz to 5.25GHz
respectively.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [ ] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [ ] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [x] I have updated the documentation (wiki pages, ReadMe etc), if applies

### CODE REVIEW

**Review by amiclaus — COMMENTED** [review]
some general comments:
- please sign the CLA
- for the documentation build - please add the documentation into the ToC (an example here: https://github.com/analogdevicesinc/no-OS/commit/510041166b2719c1d7182bdefc113db74a5767a9 )
- for the xilinx build make sure that the hardware naming in the build.json file is corresponding to the hdl output name and the project is available on main hdl branch.
- a project README.rst would be nice (see example here: https://github.com/analogdevicesinc/no-OS/commit/6673e2006bb442508840d6d90c5e50860620fca9 )

**Comment by chegbeli** [comment]
I'll remove the Xilinx platform as it is not officially supported yet.

File: drivers/frequency/adf4382/adf4382.c [hunk 1]
```diff
@@ -0,0 +1,1199 @@
+/***************************************************************************//**
+ *   @file   adf4382.c
+ *   @brief  Implementation of adf4382 Driver.
+ *   @author Ciprian Hegbeli (ciprian.hegbeli@analog.com)
+********************************************************************************
+ * Copyright 2023(c) Analog Devices, Inc.
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.c:6]: 2024 :)

File: drivers/frequency/adf4382/adf4382.c [hunk 2]
```diff
@@ -0,0 +1,1199 @@
...
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+#include "adf4382.h"
+#include "no_os_alloc.h"
+#include "no_os_error.h"
+#include "no_os_print_log.h"
+#include "no_os_delay.h"
+
+//Charge pump current values expressed in uA
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.c:46]: nit: c-style comments are preferred even for single line.

File: drivers/frequency/adf4382/adf4382.c [hunk 3]
```diff
@@ -0,0 +1,1199 @@
...
+		buff[2] = no_os_bit_swap_constant_8(ADF4382_SPI_DUMMY_DATA);
+	} else {
+		buff[0] = cmd >> 8;
+		buff[1] = cmd & 0xFF;
+		buff[2] = ADF4382_SPI_DUMMY_DATA;
+	}
+
+	ret = no_os_spi_write_and_read(dev->spi_desc, buff,
+				       ADF4382_BUFF_SIZE_BYTES);
+	if(ret < 0)
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.c:119]: `if (ret)` same as it is done for linux drivers.

File: drivers/frequency/adf4382/adf4382.c [hunk 4]
```diff
@@ -0,0 +1,1199 @@
...
+		return ret;
+
+	*data = buff[2];
+
+	return ret;
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.c:124]: `return 0;` same as it is done for linux drivers.

File: drivers/frequency/adf4382/adf4382.c [hunk 5]
```diff
@@ -0,0 +1,1199 @@
...
+	int ret;
+
+	ret = adf4382_spi_read(dev, reg_addr, &orig);
+	if (ret)
+		return ret;
+
+	tmp = orig & ~mask;
+	tmp |= data & mask;
+
+	if (tmp != orig)
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.c:148]: this check is a bit uncommon. but i'm fine with it also.

File: drivers/frequency/adf4382/adf4382.c [hunk 6]
```diff
@@ -0,0 +1,1199 @@
...
+
+/**
+ * @brief Tests if bits are set in the ADF4382 register.
+ * @param dev 	   - The device structure.
+ * @param reg_addr - The register address.
+ * @param bits 	   - Bits to be tested.
+ * @return	   - 0 in case bits are not set, positive value if they are set
+ * 		     or negative error code otherwise.
+ */
+int adf4382_spi_test_bits(struct adf4382_dev *dev, uint16_t reg_addr,
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.c:162]: is this function really needed?

File: drivers/frequency/adf4382/adf4382.c [hunk 7]
```diff
@@ -0,0 +1,1199 @@
...
+ * @param dev 	- The device structure.
+ * @return 	- 0 in case of success or negative error code.
+ */
+int adf4382_reg_dump(struct adf4382_dev *dev)
+{
+	uint8_t val;
+	uint8_t ret;
+	uint16_t i;
+
+	for(i = 0; i < 0x68; i++) {
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.c:185]: usually we use macro definitions for the registers. but this is probably similar to adf43777 and the register naming is actually the value of the register. looks good like this or macro `AD4382_REG(x)` can be used. Let's see what others prefer.

File: drivers/frequency/adf4382/adf4382.c [hunk 8]
```diff
@@ -0,0 +1,1199 @@
...
+ * it the dev structure.
+ * @param dev 		- The device structure.
+ * @param en		- The read value of the reference doubler.
+ * @return    		- 0 in case of success or negative error code.
+ */
+int adf4382_get_en_ref_doubler(struct adf4382_dev *dev, bool *en)
+{
+	int tmp;
+
+	tmp = adf4382_spi_test_bits(dev, 0x20, ADF4382_EN_RDBLR_MSK);
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.c:270]: is the "test bit" function required here. a getter function should read a register value and obtain the specific bits applying that mask. Here you don't test the bits as function header suggests. you just use the function to read the register value and apply the mask. We have for some older drivers the `read_masked` function implementation but we decided to not use it for newer drivers. so just read and apply mask here. (as in `adf4382_get_ref_div` )

File: drivers/frequency/adf4382/adf4382.c [hunk 9]
```diff
@@ -0,0 +1,1199 @@
...
+
+/**
+ * @brief Gets the user proposed output frequency
+ * @param dev 		- The device structure.
+ * @param val		- The set value of the output frequency in Hz.
+ * @return    		- Nothing
+ */
+void adf4382_get_rfout(struct adf4382_dev *dev, uint64_t *val)
+{
+	*val = dev->freq;
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.c:436]: this should be obtained from hardware.

File: drivers/frequency/adf4382/adf4382.c [hunk 10]
```diff
@@ -0,0 +1,1199 @@
...
+ * @param ch	 	- The channel to get state.
+ * @param en	 	- The status of the output channel.
+ * @return    		- 0 in case of success or negative error code.
+ */
+int adf4382_get_en_chan(struct adf4382_dev *dev, uint8_t ch, bool *en)
+{
+	int enable;
+
+	if(!ch)
+		enable = adf4382_spi_test_bits(dev, 0x2B,
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.c:526]: same here. use read function + apply mask.

File: drivers/frequency/adf4382/adf4382.c [hunk 11]
```diff
@@ -0,0 +1,1199 @@
...
+ * @brief Gets the value the sync if it is enabled or disable.
+ * @param dev 		- The device structure.
+ * @param en	 	- The read status of the sync enable.
+ * @return    		- 0 in case of success or negative error code.
+ */
+int adf4382_get_en_sync(struct adf4382_dev *dev, bool *en)
+{
+	int enable;
+
+	enable = adf4382_spi_test_bits(dev, 0x2A, ADF4382_PD_SYNC_MSK);
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.c:564]: ditto.

File: drivers/frequency/adf4382/adf4382.c [hunk 12]
```diff
@@ -0,0 +1,1199 @@
...
+	uint8_t phase_reg_value;
+	uint32_t rfout_deg_ns;
+	uint32_t phase_deg_ns;
+	uint64_t rfout_deg_s;
+	uint64_t phase_bleed;
+	uint16_t phase_deg;
+	uint64_t phase_ci;
+	uint64_t pfd_freq;
+	int ret;
+
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.c:978]: drop extra line

File: drivers/frequency/adf4382/adf4382.c [hunk 13]
```diff
@@ -0,0 +1,1199 @@
...
+		device->vco_max = ADF4382A_VCO_FREQ_MAX;
+		device->vco_min = ADF4382A_VCO_FREQ_MIN;
+		device->clkout_div_reg_val_max = ADF4382A_CLKOUT_DIV_REG_VAL_MAX;
+	}
+
+	ret = adf4382_spi_write(device, 0x00, ADF4382_RESET_CMD);
+	if (ret)
+		goto error_spi;
+
+	no_os_udelay(500000);
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.c:1138]: is this value datasheet related? if so, please add appropriate comment. and mdelay can be used instead.

File: drivers/frequency/adf4382/adf4382.c [hunk 14]
```diff
@@ -0,0 +1,1199 @@
...
+	for (clkout_div = 0; clkout_div <  dev->clkout_div_reg_val_max; clkout_div++) {
+		tmp = (1 << clkout_div) * dev->freq;
+		if (tmp < dev->vco_min || tmp > dev->vco_max )
+			continue;
+
+		vco = tmp;
+		break;
+	}
+
+	if (vco == 0) {
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.c:751]: if (vco)

File: drivers/frequency/adf4382/adf4382.h [hunk 15]
```diff
@@ -0,0 +1,721 @@
...
+#define ADF4382_CAL_VCO_TO_MSB_MSK		NO_OS_GENMASK(6, 0)
+
+/* ADF4382 REG003C Map */
+#define ADF4382_CNTR_DIV_WORD_MSK		NO_OS_GENMASK(7, 0)
+
+/* ADF4382 REG003D Map */
+#define ADF4382_SPARE_3D_MSK			NO_OS_BIT(7)
+#define ADF4382_SYNC_SP_DB_MSK			NO_OS_BIT(6)
+#define ADF4382_CMOS_OV_MSK			NO_OS_BIT(5)
+#define ADF4382_CMOS_OV(x)			no_os_field_prep(ADF4382_CMOS_OV_MSK, x)
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.h:320]: drop this and use field_prep inline.

File: drivers/frequency/adf4382/adf4382.h [hunk 16]
```diff
@@ -0,0 +1,721 @@
...
+#define ADF4382_SW_SYNC_MSK			NO_OS_BIT(7)
+#define ADF4382_SPARE_1F_MSK			NO_OS_BIT(6)
+#define ADF4382_BLEED_POL_MSK			NO_OS_BIT(5)
+#define ADF4382_EN_BLEED_MSK			NO_OS_BIT(4)
+#define ADF4382_CP_I_MSK			NO_OS_GENMASK(3, 0)
+
+/* ADF4382 REG0020 Map */
+#define ADF4382_EN_AUTOCAL_MSK			NO_OS_BIT(7)
+#define ADF4382_EN_RDBLR_MSK			NO_OS_BIT(6)
+#define ADF4382_EN_RDBLR(x)			no_os_field_prep(ADF4382_EN_RDBLR_MSK, x)
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.h:167]: drop this, use field_prep inline.

File: drivers/frequency/adf4382/adf4382.h [hunk 17]
```diff
@@ -0,0 +1,721 @@
...
+#define ADF4382_R_DIV_MSK			NO_OS_GENMASK(5, 0)
+#define ADF4382_R_DIV(x)			no_os_field_prep(ADF4382_R_DIV_MSK, x)
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.h:169]: ditto.

File: drivers/frequency/adf4382/adf4382.h [hunk 18]
```diff
@@ -0,0 +1,721 @@
...
+#define ADF4382_SPARE_53_MSK			NO_OS_BIT(7)
+#define ADF4382_PD_SYNC_MON_MSK			NO_OS_BIT(6)
+#define ADF4382_SYNC_SEL_MSK			NO_OS_BIT(5)
+#define ADF4382_RST_SYNC_MON_MSK		NO_OS_BIT(4)
+#define ADF4382_SYNC_SH_DEL_MSK			NO_OS_GENMASK(3, 0)
+
+/* ADF4382 REG0054 Map */
+#define ADF4382_ADC_ST_CNV_MSK			NO_OS_BIT(0)
+
+#define ADF4382_SPI_3W_CFG(x)			(no_os_field_prep(ADF4382_SDO_ACTIVE_MSK, x) | \
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.h:430]: ditto.

File: drivers/frequency/adf4382/adf4382.h [hunk 19]
```diff
@@ -0,0 +1,721 @@
...
+#define ADF4382_BLEED_WORD_MAX			8191
+
+#define ADF4382_PHASE_BLEED_CNST		2044000
+#define ADF4382_VCO_CAL_CNT			202
+#define ADF4382_VCO_CAL_VTUNE			124
+#define ADF4382_VCO_CAL_ALC			250
+
+#define KHZ					1000
+#define MHZ					KHZ * KHZ
+#define S_TO_NS					1000000000U
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.h:476]: same as on linux we have `no_os_units.h`. some macro definitions can be found there. if not, please add them there.

File: drivers/frequency/adf4382/adf4382.h [hunk 20]
```diff
@@ -0,0 +1,721 @@
...
+	bool				spi_3wire_en;
+	bool				cmos_3v3;
+	uint64_t			ref_freq_hz;
+	uint64_t			freq;
+	bool				ref_doubler_en;
+	uint8_t				ref_div;
+	uint8_t				cp_i;
+	uint16_t			bleed_word;
+	uint8_t				ld_count;
+	bool				adf4382a;
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.h:496]: use device id enum:
https://github.com/analogdevicesinc/no-OS/blob/87b1afc56407fc2f1790e9b1d0c7146002c84d32/drivers/frequency/adf4377/adf4377.h#L606-L610

File: drivers/frequency/adf4382/adf4382.h [hunk 21]
```diff
@@ -0,0 +1,721 @@
...
+/** ADF4382 Gets Phase adjustment polarity*/
+int adf4382_get_phase_pol(struct adf4382_dev *dev, bool *polarity);
+
+/** ADF4382 Initialization */
+int adf4382_init(struct adf4382_dev **device,
+		 struct adf4382_init_param *init_param);
+
+/** ADF4382 Remove */
+int adf4382_remove(struct adf4382_dev *dev);
+
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.h:721]: drop extra line.

File: drivers/frequency/adf4382/iio_adf4382.c [hunk 22]
```diff
@@ -0,0 +1,938 @@
+/***************************************************************************//**
+ *   @file   iio_adf4382.c
+ *   @brief  Implementation of IIO ADF4382 Driver.
+ *   @author CHegbeli (ciprian.hegbeli@analog.com)
+********************************************************************************
+ * Copyright 2023(c) Analog Devices, Inc.
```
Comment (amiclaus) [inline drivers/frequency/adf4382/iio_adf4382.c:6]: 2024

File: drivers/frequency/adf4382/adf4382.h [hunk 23]
```diff
@@ -0,0 +1,721 @@
+/***************************************************************************//**
+ *   @file   adf4382.h
+ *   @brief  Implementation of adf4382 Driver.
+ *   @author Ciprian Hegbeli (ciprian.hegbeli@analog.com)
+********************************************************************************
+ * Copyright 2023(c) Analog Devices, Inc.
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.h:6]: 2024.

File: drivers/frequency/adf4382/iio_adf4382.c [hunk 24]
```diff
@@ -0,0 +1,938 @@
...
+	uint64_t val;
+
+	if (!iio_adf4382)
+		return -EINVAL;
+
+	if (!adf4382)
+		return -EINVAL;
+	adf4382_get_ref_clk(adf4382, &val);
+	return snprintf(buf, len, "%"PRIu64, val);
+//	return 0;
```
Comment (amiclaus) [inline drivers/frequency/adf4382/iio_adf4382.c:200]: drop this.

File: drivers/frequency/adf4382/iio_adf4382.c [hunk 25]
```diff
@@ -0,0 +1,938 @@
...
+	if (!adf4382)
+		return -EINVAL;
+
+	ret = adf4382_get_en_ref_doubler(adf4382, &en);
+	if (ret)
+		return ret;
+
+	val = en;
+	return iio_format_value(buf, len, IIO_VAL_INT, 1, &val);
+
```
Comment (amiclaus) [inline drivers/frequency/adf4382/iio_adf4382.c:435]: drop extra line.

File: projects/adf4382/Makefile [hunk 26]
```diff
@@ -0,0 +1,15 @@
+# Uncomment to use the desired platform
+# PLATFORM = xilinx
+# PLATFORM = stm32
+PLATFORM = mbed
+
+# Select the example you want to enable by chossing y for enabling and n for disabling
+BASIC_EXAMPLE = n
+IIO_EXAMPLE = y
+
```
Comment (amiclaus) [inline projects/adf4382/Makefile:9]: nit: drop extra line

File: projects/adf4382/src/platform/stm32/platform_src.mk [hunk 27]
```diff
@@ -0,0 +1,16 @@
...
+	$(PLATFORM_DRIVERS)/stm32_uart.h      \
+	$(PLATFORM_DRIVERS)/stm32_uart_stdio.h
+
+SRCS += $(PLATFORM_DRIVERS)/stm32_delay.c     \
+	$(PLATFORM_DRIVERS)/stm32_gpio.c      \
+	$(PLATFORM_DRIVERS)/stm32_spi.c       \
+	$(PLATFORM_DRIVERS)/stm32_irq.c       \
+	$(PLATFORM_DRIVERS)/stm32_gpio_irq.c  \
+	$(PLATFORM_DRIVERS)/stm32_uart.c      \
+	$(PLATFORM_DRIVERS)/stm32_uart_stdio.c
```
Comment (amiclaus) [inline projects/adf4382/src/platform/stm32/platform_src.mk:16]: add new line.

**Review by amiclaus — COMMENTED** [review]
some comments on my side.

File: projects/adf4382/src/platform/xilinx/main.c [hunk 28]
```diff
@@ -0,0 +1,81 @@
...
+
+#ifdef BASIC_EXAMPLE
+	struct no_os_uart_desc *uart_desc;
+
+	ret = no_os_uart_init(&uart_desc, &adf4382_uart_ip);
+	if (ret)
+		return ret;
+
+	no_os_uart_stdio(uart_desc);
+	ret = basic_example_main();
```
Comment (rbolboac) [inline projects/adf4382/src/platform/xilinx/main.c:73]: if (ret) call no_os_uart_remove

File: projects/adf4382/src/platform/stm32/main.c [hunk 29]
```diff
@@ -0,0 +1,79 @@
...
+
+#ifdef BASIC_EXAMPLE
+	struct no_os_uart_desc *uart_desc;
+
+	ret = no_os_uart_init(&uart_desc, &adf4382_uart_ip);
+	if (ret)
+		return ret;
+
+	no_os_uart_stdio(uart_desc);
+	ret = basic_example_main();
```
Comment (rbolboac) [inline projects/adf4382/src/platform/stm32/main.c:71]: if (ret) call no_os_uart_remove

File: drivers/frequency/adf4382/adf4382.c [hunk 30]
```diff
@@ -0,0 +1,1199 @@
...
+ * @param dev	   - The device structure.
+ * @param reg_addr - The register address.
+ * @param data 	   - Data value to write.
+ * @return 	   - 0 in case of success or negative error code otherwise.
+ */
+int adf4382_spi_write(struct adf4382_dev *dev, uint16_t reg_addr, uint8_t data)
+{
+	uint8_t buff[ADF4382_BUFF_SIZE_BYTES];
+	uint16_t cmd;
+
```
Comment (rbolboac) [inline drivers/frequency/adf4382/adf4382.c:77]: dev should be checked for NULL

File: drivers/frequency/adf4382/adf4382.c [hunk 31]
```diff
@@ -0,0 +1,1199 @@
...
+ * @param reg_addr - The register address.
+ * @param data	   - Data read from the device.
+ * @return	   - 0 in case of success or negative error code otherwise.
+ */
+int adf4382_spi_read(struct adf4382_dev *dev, uint16_t reg_addr, uint8_t *data)
+{
+	int ret;
+	uint8_t buff[ADF4382_BUFF_SIZE_BYTES];
+	uint16_t cmd;
+
```
Comment (rbolboac) [inline drivers/frequency/adf4382/adf4382.c:105]: dev should be checked for NULL

File: drivers/frequency/adf4382/adf4382.c [hunk 32]
```diff
@@ -0,0 +1,1199 @@
...
+ * @brief Set the desired reference frequency and reset everything over to maximum
+ * supported value of 5GHz to the max. value and everything under the minimum
+ * supported value of 10MHz to the min. value.
+ * @param dev 		- The device structure.
+ * @param val		- The desired reference frequency in Hz.
+ * @return    		- 0 in case of success or negative error code.
+ */
+int adf4382_set_ref_clk(struct adf4382_dev *dev, uint64_t val)
+{
+	dev->ref_freq_hz = (uint64_t)val;
```
Comment (rbolboac) [inline drivers/frequency/adf4382/adf4382.c:222]: does not look like the cast is needed

File: drivers/frequency/adf4382/adf4382.c [hunk 33]
```diff
@@ -0,0 +1,1199 @@
...
+
+	while (y != 0) {
+		tmp = x % y;
+		x = y;
+		y = tmp;
+	}
+
+	return x;
+}
+
```
Comment (rbolboac) [inline drivers/frequency/adf4382/adf4382.c:607]: should be moved in no_os_util or use the implementation already available there

File: drivers/frequency/adf4382/adf4382.c [hunk 34]
```diff
@@ -0,0 +1,1199 @@
...
+	//The charge pump current will also need to be taken in to account
+	phase_ci = phase_bleed * adf4382_ci_ua[dev->cp_i];
+	phase_ci = no_os_div_u64(phase_ci, UA_TO_A);
+
+	//Computation of the register value for the phase adjust
+	pfd_freq = adf4382_pfd_compute(dev);
+	phase_reg_value = no_os_div_u64((phase_ci * pfd_freq), (360 * dev->freq));
+
+	if(phase_reg_value > 255)
+		phase_reg_value -= 255;
```
Comment (rbolboac) [inline drivers/frequency/adf4382/adf4382.c:1020]: phase_reg_value maximum value is 255, this will never happen. did you mean == instead of > ?, otherwise please remove lines

File: drivers/frequency/adf4382/adf4382.c [hunk 35]
```diff
@@ -0,0 +1,1199 @@
...
+ * @return    		- Result of the tesint procedure, negative error code
+ * 			  otherwise.
+ */
+int adf4382_get_phase_pol(struct adf4382_dev *dev, bool *polarity)
+{
+	uint8_t pol;
+
+	pol = adf4382_spi_test_bits(dev, 0x32, ADF4382_PHASE_ADJ_POL_MSK);
+	if (pol < 0)
+		return pol;
```
Comment (rbolboac) [inline drivers/frequency/adf4382/adf4382.c:1058]: this will never be true, pol is uint8_t, always >= 0, please adapt accordingly

File: drivers/frequency/adf4382/adf4382.c (hunk 6 above)
Comment (rbolboac) [inline drivers/frequency/adf4382/adf4382.c:162]: the return type should be the execution status of the API. pass the result by pointer in the parameters list

File: drivers/frequency/adf4382/iio_adf4382.c [hunk 36]
```diff
@@ -0,0 +1,938 @@
...
+ * @param dev	  - The iio device structure.
+ * @param reg	  - Address of the register to be read from.
+ * @param readval - Read data.
+ * @return	  - Result of the reading procedure.
+ */
+static int adf4382_iio_read_reg(struct adf4382_iio_dev *dev, uint32_t reg,
+				uint32_t *readval)
+{
+	return adf4382_spi_read(dev->adf4382_dev, (uint16_t)reg,
+				(uint8_t *)readval);
```
Comment (rbolboac) [inline drivers/frequency/adf4382/iio_adf4382.c:103]: casting this pointer will result in a different value based on the endianness of the CPU. please do not make the cast

File: drivers/frequency/adf4382/iio_adf4382.c [hunk 37]
```diff
@@ -0,0 +1,938 @@
...
+ * @param priv    - Command attribute id.
+ * @return 	  - The size of the read data in case of success, error code
+ * 		    otherwise.
+ */
+static int adf4382_iio_read_freq(void *dev, char *buf, uint32_t len,
+				 const struct iio_ch_info *channel,
+				 intptr_t priv)
+{
+	struct adf4382_iio_dev *iio_adf4382 = (struct adf4382_iio_dev *)dev;
+	struct adf4382_dev *adf4382 = iio_adf4382->adf4382_dev;
```
Comment (rbolboac) [inline drivers/frequency/adf4382/iio_adf4382.c:135]: this should be done after the null check of iio_adf4382, please see all occurrences in the file

File: drivers/frequency/adf4382/adf4382.c (hunk 14 above)
Comment (chegbeli) [reply drivers/frequency/adf4382/adf4382.c:751]: if (!vco)

File: drivers/frequency/adf4382/adf4382.c (hunk 9 above)
Comment (chegbeli) [reply drivers/frequency/adf4382/adf4382.c:436]: The value is not readily available from the registers and would need to be recalculated from multiple registers. this would require several multiplication and division operations and would loose precision along the way.

File: drivers/frequency/adf4382/adf4382.c (hunk 5 above)
Comment (chegbeli) [reply drivers/frequency/adf4382/adf4382.c:148]: adapted the function from the Linux implementation https://elixir.bootlin.com/linux/latest/source/drivers/base/regmap/regmap.c#L3137

**Comment by chegbeli** [comment]
V1 -> V2
* Removed Xilinx platform, as there is no official support for it
* Updated sdp-ck1z.ioc to use version 6.5.0
* Added missing driver and project documentation
* Fixed wrong POR delay time
* Removed warning messages as they brake the IIO over UART
* Remove adf4382_spi_test_bits and replaced with no_os_field_get
* Added adf4382_get_rfout which computes the current output frequency from register values
* Replaced some MACROs with the ones from Use no_os_units.h
* Removed ADF4382_EN_RDBLR(x), ADF4382_R_DIV(x) and ADF4382_CMOS_OV(x) MACROs
* Added NULL checks to API functions
* Moved dev NULL checks before attributing the dev in the IIO functions
* Fixed pointer casting from u32 to u8
* Added ID structure entries for ADF4382 and ADF4382A
* Added missing no_os_uart_remove to the main files
* Fixed missing adf4382_remove for error conditions in the IIO and basic example

**Review by amiclaus — COMMENTED** [review]
there are some conflicts that need addressing.

File: drivers/frequency/adf4382/adf4382.c (hunk 7 above)
Comment (buha) [reply drivers/frequency/adf4382/adf4382.c:171]: This is a debugging only function that users wouldn't not use in "production". I wouldn't care much about that. It's probably even easier to follow the datasheet as Cipi wrote it because you normally have a large table with register names and addresses and you see gaps/discontinuities in addresses right away when you look at the addresses.

File: drivers/frequency/adf4382/adf4382.c [hunk 38]
```diff
@@ -0,0 +1,1303 @@
...
+	phase_deg_ns = rfout_deg_ns * phase_ps;
+	//Convert it to degrees/ps
+	phase_deg = no_os_div_u64(phase_deg_ns, NS_TO_PS);
+
+	if(phase_deg > 360) {
+		pr_err("Phase Adjustment cannot exceed 360deg per Clock Period\n");
+		return EINVAL;
+	}
+
+	/*Phase adjustment can only be done if bleed is active, and a bleed
```
Comment (buha) [inline drivers/frequency/adf4382/adf4382.c:1104]: there is some inconsistency here in how you wrote comments with // or with /*, please make them look alike

File: drivers/frequency/adf4382/adf4382.c [hunk 39]
```diff
@@ -0,0 +1,1303 @@
...
+		break;
+	case ID_ADF4382A:
+		device->freq_max = ADF4382A_RFOUT_MAX;
+		device->freq_min = ADF4382A_RFOUT_MIN;
+		device->vco_max = ADF4382A_VCO_FREQ_MAX;
+		device->vco_min = ADF4382A_VCO_FREQ_MIN;
+		device->clkout_div_reg_val_max = ADF4382A_CLKOUT_DIV_REG_VAL_MAX;
+		break;
+	default:
+		return -EINVAL;
```
Comment (buha) [inline drivers/frequency/adf4382/adf4382.c:1234]: you've already allocated something, you'll need to `goto` to a label to avoid memory leaks

**Review by buha — COMMENTED** [review]
Looks good, some minor remarks above.
One more thing, Ramona commented about the gcd() function, (like her) I think you should use the util.h one and remove your implementation unless there is a justification for keeping it which would go in a code comment.

**Comment by chegbeli** [comment]
> Looks good, some minor remarks above. One more thing, Ramona commented about the gcd() function, (like her) I think you should use the util.h one and remove your implementation unless there is a justification for keeping it which would go in a code comment.

Unfortunately the function in no_os_util.h has a slow execution time which is visible in this driver due to the large numbers which are being processed. I could make a new PR for the gcd replacing the one we currently have, if that is something that would help us.

File: projects/adf4382/src/platform/stm32/platform_src.mk [hunk 40]
```diff
@@ -0,0 +1,17 @@
...
+
```
Comment (rbolboac) [inline projects/adf4382/src/platform/stm32/platform_src.mk:17]: missing new line at the end of file

**Comment by tfcollins** [comment]
It doesn't seem like the driver checks for the PLL actually locking. REG 0x58 (bit 0). After initialization it should be the final check.

File: drivers/frequency/adf4382/adf4382.c [hunk 41]
```diff
@@ -0,0 +1,1303 @@
...
+
+/**
+ * @brief Will output on the terminal the values of all the ADF4382 registers.
+ * @param dev 	- The device structure.
+ * @return 	- 0 in case of success or negative error code.
+ */
+int adf4382_reg_dump(struct adf4382_dev *dev)
+{
+	uint8_t val;
+	uint8_t ret;
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.c:168]: if ret can take negative values, then use `int` type. Check throughout the code, since this is occuring multiple times.

File: drivers/frequency/adf4382/adf4382.c [hunk 42]
```diff
@@ -0,0 +1,1303 @@
...
+		buff[2] = no_os_bit_swap_constant_8(ADF4382_SPI_DUMMY_DATA);
+	} else {
+		buff[0] = cmd >> 8;
+		buff[1] = cmd & 0xFF;
+		buff[2] = ADF4382_SPI_DUMMY_DATA;
+	}
+
+	ret = no_os_spi_write_and_read(dev->spi_desc, buff,
+				       ADF4382_BUFF_SIZE_BYTES);
+	if(ret)
```
Comment (amiclaus) [inline drivers/frequency/adf4382/adf4382.c:125]: please follow the linux guidelines -> `if (ret)` Check the entire code since there are multiple occurences. Same applies for other statements (`for`, `while`, `switch` etc).

**Comment by chegbeli** [comment]
V2 -> V3
*Fixed coding style
*Updated return variables to proper data types
*Add locked check
*Switched to no_os_greatest_common_divisor
*Fixed not freeing memory when failing checks

File: drivers/frequency/adf4382/iio_adf4382.c [hunk 43]
```diff
@@ -0,0 +1,989 @@
...
+/**
+ * @brief Helper function to read a pair of values in a table with 2 elements.
+ * @param tb	- Input table.
+ * @param size	- Size of the table.
+ * @param val1	- The first value to be found.
+ * @param val2	- The second value to be found.
+ * @return i	- The index of the pair if found or an error if there is no
+ * 		  pair of values in the table.
+ */
+static int adf4382_iio_find_2d_row(const int (*tbl)[2], const int size,
```
Comment (CiprianRegus) [inline drivers/frequency/adf4382/iio_adf4382.c:79]: You may use `ci_table` directly in this function without needing to pass it as a parameter.

**Review by CiprianRegus — APPROVED** [review]
